OSDOCS#8867: Adding failure domains to Nutanix docs (#69618)
openshift/openshift-docs: modules/installation-configuring-nutanix-failure-domains.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-nutanix-failure-domains_{context}"]
= Configuring failure domains

Failure domains improve the fault tolerance of an {product-title} cluster by distributing control plane and compute machines across multiple Nutanix Prism Elements (clusters).

[TIP]
====
It is recommended that three failure domains be configured to ensure high-availability.
====

.Prerequisites

* You have an installation configuration file (`install-config.yaml`).

.Procedure

. Edit the `install-config.yaml` file and add the following stanza to configure the first failure domain:
+
[source,yaml]
----
# ...
platform:
  nutanix:
    failureDomains:
    - name: <failure_domain_name>
      prismElement:
        name: <prism_element_name>
        uuid: <uuid>
      subnetUUIDs:
      - <network_uuid>
# ...
----
+
where:

`<failure_domain_name>`:: Specifies a unique name for the failure domain. The name is limited to 64 or fewer characters, which can include lower-case letters, digits, and a dash (`-`). The dash cannot be in the leading or ending position of the name.

> `<prism_element_name>`:: Optional. Specifies the name of the Prism Element.

Reviewer: maybe add more details on prismElement
is it name or uuid ? what is mandatory ? which takes precedence ?

Author: First- nice to meet you @tuxtof. Yanahua had commented on this separately. At the time you commented, I had forgotten to add a description for `<prism_element_uuid>`. Does this help to clarify things?

Reviewer: Hey , nice to meet you too
you specify the name as optional but not the uuid
can we specify only the name ? if not what interest to specify name if uuid is mandatory ?

// Module included in the following assemblies:
//
// * installing/installing_nutanix/installing-nutanix-installer-provisioned.adoc
// * installing/installing_nutanix/installing-restricted-networks-nutanix-installer-provisioned.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-nutanix-failure-domains_{context}"]
= Configuring failure domains

Failure domains improve the fault tolerance of an {product-title} cluster by distributing control plane and compute machines across multiple Nutanix Prism Elements (clusters).

.Prerequisites

* You have an installation configuration file (`install-config.yaml`).

.Procedure

. Edit the `install-config.yaml` file and add the following stanza to configure the first failure domain:
+
[source,yaml]
----
# ...
platform:
  nutanix:
    failureDomains:
    - name: <failure_domain_name>
      prismElement:
        name: <prism_element_name>
        uuid: <uuid>

>       endpoint:

Reviewer 2: Remove the "endpoint" section, the failureDomain does not have the "endpoint" field.

>         address: <address>

Reviewer 2: Remove the "endpoint" section, the failureDomain does not have the "endpoint" field.

>         address: <address>

Reviewer 2: Remove the "endpoint" section, the failureDomain does not have the "endpoint" field.

Author: Done.

      prismElement:
        name: <prism_element_name>

>         uuid: <prism_element_uuid>

Reviewer 3: @yanhua121 That's a little strange to me, so should we remove `endpoint` from above field explanation?

>         uuid: <prism_element_uuid>

Reviewer 2: @sgaoshang @mjpytlak Since the endpoint of the prismElement is optional, and here in the use case of failureDomain, the PE endpoint is not used by the installer so we don't bother to document it. But in the use case of platform.nutanix.prismElements, the PE endpoint is used by the installer to construct the Infrastructure CR manifest. But if you feel we need to document all the optional fields, no matter it is used or not, @mjpytlak You can add the endpoint documentation here and mark it is optional.

>         uuid: <prism_element_uuid>

Reviewer 3: @yanhua121 Thanks for the explanation, I'm OK with both way then.

# ...
----
+
where:

`<failure_domain_name>`:: Specifies a name for the failure domain.
`<prism_element_name>`:: Optional. Specifies the name of the Prism Element.
`<uuid>`:: Specifies the universally unique identifier (UUID) of the Prism Element.
`<address>`:: Optional. Specifies the domain name or IP address of the Prism Element.

> `<port>`:: Optional. Specifies the port that is used to log into Prism Element.

Reviewer 2: Remove.

Author: Done.

> `<network_uuid`:: Specifies the UUID of the Prism Element network that contains the virtual IP addresses and DNS records that you configured. Only one subnet per {product-title} cluster is supported.

Reviewer 2: Only one subnet per Failure Domain is supported.

Author: Done.

. As required, configure additional failure domains.
. Do one of the following:

** If compute and control plane machines can share the same set of failure domains, list the failure domain names under the cluster's default machine configuration.
+
.Example of control plane and compute machines sharing a set of failure domains
+
[source,yaml]
----
# ...
platform:
  nutanix:
    defaultMachinePlatform:
      failureDomains:
        - failure-domain-1
        - failure-domain-2
        - failure-domain-3
# ...
----
** If compute and control plane machines must use different failure domains, list the failure domain names under the respective machine pools.
+
.Example of control plane and compute machines using different failure domains
+
[source,yaml]
----
# ...
controlPlane:
  nutanix:
    failureDomains:
      - failure-domain-1
      - failure-domain-2
      - failure-domain-3
# ...
compute:
  nutanix:
    failureDomains:
      - failure-domain-1
      - failure-domain-2
# ...

> ----

Reviewer 3: failure domains under the respective machine pools should be under `platform` like
```
controlPlane:
  platform:
    nutanix:
      failureDomains:
        - failure-domain-1
        - failure-domain-2
        - failure-domain-3
compute:
  platform:
    nutanix:
      failureDomains:
        - failure-domain-1
        - failure-domain-2
```

Author: Good catch. Done.

// Module included in the following assemblies:
//
// * installing/installing_nutanix/installing-nutanix-installer-provisioned.adoc
// * installing/installing_nutanix/installing-restricted-networks-nutanix-installer-provisioned.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-nutanix-failure-domains_{context}"]
= Configuring failure domains

Failure domains improve the fault tolerance of an {product-title} cluster by distributing control plane and compute machines across multiple Nutanix Prism Elements (clusters).

[TIP]
====

> It is recommended that three failure domains be configured to ensure high-availability.

Reviewer 4: ```suggestion
It is recommended that you configure three failure domains to ensure high-availability.
```

Author: Done. Thank you.

====

.Prerequisites

* You have an installation configuration file (`install-config.yaml`).

.Procedure

. Edit the `install-config.yaml` file and add the following stanza to configure the first failure domain:
+
[source,yaml]

> ----

Reviewer 4: Not sure if it applies to install-config, but per the docs guidelines, we are supposed to include [the metadata stuff](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#source-tags-for-terminal-commands-and-output) (at least the api and name, I think). I defer to your expertise with the install-configs.

Author: I am not sure if it applies to `install-config` but I do not think it can hurt. I added the first three lines of a typical file, which includes the api info.

# ...

> platform:

Reviewer 4: Can the user configure different failure domains for control plane and worker nodes here (as shown in the following step)?

  nutanix:
    failureDomains:
    - name: <failure_domain_name>
      prismElement:
        name: <prism_element_name>
        uuid: <prism_element_uuid>
      subnetUUIDs:
      - <network_uuid>
# ...
----
+
where:

`<failure_domain_name>`:: Specifies a unique name for the failure domain. The name is limited to 64 or fewer characters, which can include lower-case letters, digits, and a dash (`-`). The dash cannot be in the leading or ending position of the name.
`<prism_element_name>`:: Optional. Specifies the name of the Prism Element.
`<prism_element_uuid`>:: Specifies the UUID of the Prism Element.
`<network_uuid`>:: Specifies the UUID of the Prism Element subnet object. The subnet's IP address prefix (CIDR) should contain the virtual IP addresses that the {product-title} cluster uses. Only one subnet per failure domain (Prism Element) in an {product-title} cluster is supported.

> . As required, configure additional failure domains.

Reviewer 4: ```suggestion
. As required, configure additional failure domains by performing one of the following steps:
```

> . Do one of the following:

Reviewer 4: ```suggestion

```

Author: Yes. We do need both steps here. After you configure the first failure domain, you configure subsequent failure domains. "As required, configure additional failure domains".

After which, you either add the failure domains to the default machine configuration or you can add specific failure domains to the control plane and compute machine pools.

Reviewer 4: @mjpytlak Sorry. I wasn't sure how to present this. I am thinking we don't need these two steps. 

```
. As required, configure additional failure domains.
. Do one of the following:
```
Maybe combine into one?

`. As required, configure additional failure domains by doing the following:`

Author: @mburke5678 These are, in fact, two distinct steps that should not be combined. I am happy to jump on a call with you to explain. But as the procedure stands, it is technically accurate and technically complete.

Reviewer 5: Wonder if something describing why they do this would help? Like:
```suggestion
. To use the configured failure domains, do one of the following:
```

I have no idea if that's the right thing to say, but just to illustrate the concept

Author: Thanks @jeana-redhat I went with "To distribute control plane and compute machines across the failure domains, do one of the following:"

This is language we use elsewhere in the docs.

> ** If compute and control plane machines can share the same set of failure domains, list the failure domain names under the cluster's default machine configuration.

Reviewer 4: ```suggestion
** If compute and control plane machines can share the same set of failure domains, add the failure domain names under the cluster's default machine configuration.
```

Author: Done. Thank you.

+
.Example of control plane and compute machines sharing a set of failure domains
+
[source,yaml]
----
# ...
platform:
  nutanix:
    defaultMachinePlatform:
      failureDomains:
        - failure-domain-1
        - failure-domain-2
        - failure-domain-3
# ...
----

> ** If compute and control plane machines must use different failure domains, list the failure domain names under the respective machine pools.

Reviewer 4: ```suggestion
** If compute and control plane machines must use different failure domains, add the failure domain names under the respective machine pools.
```

Author: Done. Thank you.
